### PR TITLE
[#470] Migrate to Kover 0.7.0

### LINKED ISSUE #470: Migrate to Kover 0.7.0
## Why

[Kover 0.7.0](https://github.com/Kotlin/kotlinx-kover/releases/tag/v0.7.0) brings many new updates also `breaking changes` that require us to migrate the legacy config to the latest. Please follow this migration guide https://github.com/Kotlin/kotlinx-kover/blob/v0.7.0/docs/gradle-plugin/migrations/migration-to-0.7.0.md for the update.

## Who Benefits?

Developers

### PR DESCRIPTION
**Pull request by kaungkhantsoe**
- Closes #470 

## What happened 👀

[Kover 0.7.0](https://github.com/Kotlin/kotlinx-kover/releases/tag/v0.7.0) brings many new updates also `breaking changes` that require us to migrate the legacy config to the latest. Please follow this migration guide https://github.com/Kotlin/kotlinx-kover/blob/v0.7.0/docs/gradle-plugin/migrations/migration-to-0.7.0.md for the update.

## Insight 📝

- We can no longer generate both `xml` and `html` reports with one command like `koverMergedReport`. 
- So we need to run 2 commands, one to generate `xml` report and another one to generate `html` report for each `koverMergedReport` command
- In order to make a merged report, we need to declare the `Kover Configs` inside [App level build.gradle](https://github.com/Kotlin/kotlinx-kover/blob/main/kover-gradle-plugin/examples/android/multiproject/app/build.gradle.kts)

## Proof Of Work 📹

| Type | Before | After |
| --- | --- | --- | 
| Template Compose | Old Compose | New Compose |
| Template Xml | Old XML | New XML |

### CODE REVIEW

File: sample-compose/README.md [hunk 1]
```diff
@@ -38,10 +38,10 @@ $ ./gradlew domain:test
 - Run unit testing with coverage:
 
 ```
-$ ./gradlew koverMergedHtmlReport
+$ ./gradlew koverHtmlReport
 ```
 
-Report is located at: `./build/reports/kover/merged/`
+Report is located at: `.app/build/reports/kover/`
```
Comment (luongvo) [inline sample-compose/README.md:44]: No `.` before app

```suggestion
Report is located at: `app/build/reports/kover/`
```

File: sample-compose/app/build.gradle.kts [hunk 2]
```diff
@@ -108,8 +108,10 @@ android {
...
             if (it.name != "testStagingDebugUnitTest") {
-                it.extensions.configure(kotlinx.kover.api.KoverTaskExtension::class) {
-                    isDisabled.set(true)
+                kover {
+                    excludeTests {
+                        tasks(it.name)
+                    }
                 }
             }
         }
```
Comment (luongvo) [inline sample-compose/app/build.gradle.kts:117]: [This approach is deprecated](https://github.com/Kotlin/kotlinx-kover/blob/b35990f1a922cb793977b9627f2a7586fa3533bc/README.md#configuring-jvm-test-task) and has been replaced with `mergeWith("stagingDebug")` from the default `koverReport` config as below.

File: sample-compose/app/build.gradle.kts [hunk 3]
```diff
@@ -177,3 +179,46 @@ dependencies {
...
+
+dependencies {
+    kover(project(":data"))
+    kover(project(":domain"))
+}
+
+koverReport {
+    defaults {
+        mergeWith("stagingDebug")
+        mergeWith("stagingRelease")
```
Comment (luongvo) [inline sample-compose/app/build.gradle.kts:195]: I think, by default, the project should only support `debug` flavor, as we do not often have custom code in `release` flavor. Merging with `release` flavor will duplicate this operation.

image


```suggestion
        mergeWith("stagingDebug")
```

File: sample-compose/data/build.gradle.kts [hunk 4]
```diff
@@ -45,8 +45,10 @@ android {
...
             if (it.name != "testDebugUnitTest") {
-                it.extensions.configure(kotlinx.kover.api.KoverTaskExtension::class) {
-                    isDisabled.set(true)
+                kover {
+                    excludeTests {
+                        tasks(it.name)
+                    }
                 }
             }
         }
```
Comment (luongvo) [inline sample-compose/data/build.gradle.kts:54]: same, [this approach is deprecated](https://github.com/Kotlin/kotlinx-kover/blob/b35990f1a922cb793977b9627f2a7586fa3533bc/README.md#configuring-jvm-test-task). Using `mergeWith("stagingDebug")` will trigger the `data:testDebugUnitTest`.

File: .cicdtemplate/.codemagic/codemagic.yaml [hunk 5]
```diff
@@ -22,9 +22,11 @@ definitions:
       script: ./gradlew detekt
     - &unit_test
       name: Run unit tests
-      script: ./gradlew koverMergedReport
+      script: |
+        ./gradlew koverHtmlReport
+        ./gradlew koverXmlReport
```
Comment (luongvo) [inline .cicdtemplate/.codemagic/codemagic.yaml:27]: XML report is for Danger to execute. Do we need it in this workflow?

File: Dangerfile [hunk 6]
```diff
@@ -32,10 +32,10 @@ end
 # Show Danger test coverage report from Kover for templates
 # Report coverage of modified files, warn if total project coverage is under 80%
 # or if any modified file's coverage is under 90%
```
Comment (luongvo) [inline Dangerfile:34]: The comment should match to the code 😮‍💨 

```suggestion
# or if any modified file's coverage is under 95%
```

File: codemagic.yaml [hunk 7]
```diff
@@ -21,14 +21,18 @@ definitions:
     - &unit_test_on_template_compose
       name: Run unit tests on template-compose
       working_directory: ./template-compose
-      script: ./gradlew koverMergedReport
+      script: |
+        ./gradlew koverHtmlReport
+        ./gradlew koverXmlReport
```
Comment (luongvo) [inline codemagic.yaml:26]: same

File: codemagic.yaml [hunk 8]
```diff
@@ -21,14 +21,18 @@ definitions:
...
     - &unit_test_on_template_xml
       name: Run unit tests on template-xml
       working_directory: ./template-xml
-      script: ./gradlew koverMergedReport
+      script: |
+        ./gradlew koverHtmlReport
+        ./gradlew koverXmlReport
```
Comment (luongvo) [inline codemagic.yaml:32]: ditto here

**Review by luongvo — CHANGES_REQUESTED** [review]
@kaungkhantsoe I reviewed all necessary config in the sample-compose and workflows first.

File: .cicdtemplate/.codemagic/codemagic.yaml (hunk 5 above)
Comment (kaungkhantsoe) [reply .cicdtemplate/.codemagic/codemagic.yaml:27]: You are right @luongvo . I don't see any task running Danger in the workflow. I am not sure why we need `koverMergedReport` initially. I just migrated to the closes thing I can 😆 . Shall we only run HTML report then? What do you think @lydiasama? Do you have any concerns about it?

Comment (lydiasama) [reply .cicdtemplate/.codemagic/codemagic.yaml:27]: I just followed the previous workflow when initialing this codemagic workflow. 🤔 
BTW, I don’t have any concerns.  Let’s remove the XML report if we don’t need it 😬  @kaungkhantsoe  @luongvo

**Comment by kaungkhantsoe** [comment]
@luongvo @lydiasama All updated 🙏

**Review by luongvo — APPROVED** [review]
GJ 💯

**Review by ryan-conway — APPROVED** [review]

**Review by doannimble — APPROVED** [review]

**Pull request merged**